FastSimulation/Calorimetry/src/CaloResponse.cc

```diff
@@ -0,0 +1,82 @@
+#include "FastSimulation/Calorimetry/interface/CaloResponse.h"
+
+CaloResponse::CaloResponse( const edm::ParameterSet& pset )
+{
+
+  // Get the filename and histogram name
+  std::string ecalScalesFileName_ = pset.getUntrackedParameter<std::string>("fileName");
+  std::string ecalScalesHistogramName_ = pset.getUntrackedParameter<std::string>("histogramName");
+
+  // Read histo
+  edm::FileInPath myDataFile( ecalScalesFileName_ ); // TODO: check if this is necessary
+  TFile * myFile = new TFile( myDataFile.fullPath().c_str(), "READ" );
+
+  gROOT->cd(); // create histogram in memory
+  auto obj = myFile->Get( ecalScalesHistogramName_.c_str() );
+
+  // Check if histogram exists in file
+  if(!obj) {
+    throw cms::Exception( "FileReadError" )
+      << "Histogram \"" << ecalScalesHistogramName_
+      << "\" not found in file \"" << ecalScalesFileName_
+      << "\", used for correcting the response of the ECAL in FastSim.\n";
+  }
+  h3_ = new TH3F( *((TH3F*)obj) );
+
+  myFile->Close();
+  delete myFile;
+
+}
+
+
+float CaloResponse::getScale( const RawParticle& particleAtEcalEntrance,
+    const std::map<CaloHitID,float>& hitMap ) const
+{
+
+  double simE = 0; // total simulated energy for this particle
+  for( auto mapIterator : hitMap ) {
+    simE += mapIterator.second;
+  }
+
+  float genEta = std::abs( particleAtEcalEntrance.eta() );
+  float genE = particleAtEcalEntrance.e();
+
+  return getScale( genE, genEta, simE );
+
+}
+
+
+float CaloResponse::getScale( float genE, float genEta, float simE ) const
+{
+
+  float r = simE / genE;
+
+  int binE   = h3_->GetXaxis()->FindFixBin( genE );
+  int binEta = h3_->GetYaxis()->FindFixBin( genEta );
+
+  // find the two bins which are closest to the actual value
```

Comment:
Please have a look at slide 7 of https://indico.cern.ch/event/434525/session/0/contribution/9/attachments/1130590/1615842/2015-05-newMethod.pdf
The interpolation in the variable you are looking at (in this case r) is important for large gradients in the function. An interpolation in energy/η is not studied yet.